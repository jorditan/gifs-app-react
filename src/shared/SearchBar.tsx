import { useEffect, useState } from "react";
import { Button } from "@/components/ui/button";
import { Input } from "@/components/ui/input";
import { Search } from "lucide-react";
import { ButtonIcon } from "./ButtonIcon";
interface Props {
  placeholder: string
  page: number;
  reset: boolean,
  onQuery: (query: string) => void;
}

export const SearchBar = ({ placeholder, onQuery, reset }: Props) => {
  const [query, setQuery] = useState('');

  useEffect(() => {
    const timeoutId = setTimeout(() => {
      onQuery(query)
    }, 700);

    return () => {
      clearTimeout(timeoutId);
    }

  }, [query, onQuery])

  const handleSearch = () => {
    onQuery(query);
    if (reset) {
      setQuery('')
    }
  }

  const handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
    if (event.key === 'Enter') {
      handleSearch();
    }
  }

  return (
    <>
      <div className="search-container flex gap-3">
        <Input type="text"
          className="text-[#fafafa] w-full"
          value={query}
          onChange={(event) => setQuery(event.target.value)}
          placeholder={placeholder}
          onKeyDown={handleKeyDown}>
        </Input>
        <Button
          variant="secondary"
          onClick={handleSearch}
          className="hidden md:inline-flex"
        >
          Buscar
        </Button>
        <ButtonIcon
          tooltipText="Buscar"
          icon={Search}
          variant="secondary"
          handleAction={handleSearch}
          className="md:hidden"
        >
        </ButtonIcon>
      </div>
    </>
  )
}
