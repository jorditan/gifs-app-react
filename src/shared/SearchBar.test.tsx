import { describe, expect, test, vi } from "vitest"
import { SearchBar } from "./SearchBar"
import { fireEvent, render, screen, waitFor } from "@testing-library/react"

describe('SearchBar', () => {
  test('should render searchbar correctly', () => {
    const { container } = render(<SearchBar onQuery={() => { }} placeholder={""} />)

    expect(container).toMatchSnapshot()
    expect(screen.getByRole('textbox')).toBeDefined();
    expect(screen.getByRole('button')).toBeDefined();
  })

  test('should call onQuery with the correct value after 700ms', async () => {
    const onQuery = vi.fn();
    render(<SearchBar onQuery={onQuery} placeholder={""} />)

    const input = screen.getByRole('textbox');
    fireEvent.change(input, { target: { value: 'test' } });

    await waitFor(() => {
      expect(onQuery).toHaveBeenCalledWith('test')
    })
  })

  test('should call only once with the las value (debounce)', async () => {
    const onQuery = vi.fn();
    render(<SearchBar onQuery={onQuery} placeholder={""} />)

    const input = screen.getByRole('textbox');
    fireEvent.change(input, { target: { value: 't' } });
    fireEvent.change(input, { target: { value: 'te' } });
    fireEvent.change(input, { target: { value: 'tes' } });
    fireEvent.change(input, { target: { value: 'test' } });

    await waitFor(() => {
      expect(onQuery).toHaveBeenCalledWith('test')
      expect(onQuery).toHaveBeenCalledTimes(1)
    })
  })

  test('should call onQuery when button clicked with the input value', async () => {
    const onQuery = vi.fn();
    render(<SearchBar onQuery={onQuery} placeholder={""} />)

    const button = screen.getByRole('button');
    const input = screen.getByRole('textbox');

    fireEvent.change(input, { target: { value: 'test' } });
    fireEvent.change(button);

    await waitFor(() => {
      expect(onQuery).toHaveBeenCalledTimes(1);
      expect(onQuery).toHaveBeenCalledWith('test')
    })
  })

  test('should the input has the correct placeholder value', () => {
    const placeholderValue = "buscar";
    render(<SearchBar onQuery={() => { }} placeholder={placeholderValue} />)

    expect(screen.getByPlaceholderText(placeholderValue));
  })

  test('should call onQuery when key "Enter" is called', () => {
    const onQuery = vi.fn();
    render(<SearchBar onQuery={onQuery} placeholder={""} />)

    const input = screen.getByRole('textbox');
    fireEvent.change(input, { target: { value: 'test' } });

    fireEvent.keyDown(input, { key: 'Enter', code: 'Enter', charCode: 13 });

    expect(onQuery).toHaveBeenCalledTimes(1);
  })
})