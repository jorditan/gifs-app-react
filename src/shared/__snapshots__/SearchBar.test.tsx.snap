// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchBar > should render searchbar correctly 1`] = `
<div>
  <div
    class="search-container flex gap-3"
  >
    <input
      class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input h-9 min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive text-[#fafafa] w-full"
      data-slot="input"
      placeholder=""
      type="text"
      value=""
    />
    <button
      class="items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-secondary text-secondary-foreground hover:bg-secondary/80 h-9 px-4 py-2 has-[>svg]:px-3 hidden md:inline-flex"
      data-slot="button"
    >
      Buscar
    </button>
    <button
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-secondary text-secondary-foreground hover:bg-secondary/80 size-9 md:hidden"
      data-slot="tooltip-trigger"
      data-state="closed"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-search"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m21 21-4.34-4.34"
        />
        <circle
          cx="11"
          cy="11"
          r="8"
        />
      </svg>
    </button>
  </div>
</div>
`;
